_posts/2020-05-18-websocket-fallback-sockjs.md

```diff
@@ -0,0 +1,329 @@
++           .withSockJS();
+    }
+}
+{% endhighlight %}
+
+* web-socket-manager.ts
+{% highlight diff %}
++import * as SockJS from 'sockjs-client';
++
+ export class WebSocketManager {
+    private webSocket?: WebSocket;
+    constructor() {
+        this.webSocket = this.initWebSocket();
+    }
+    
+    initWebSocket(): WebSocket {
+-      return new WebSocket('https://localhost/signaling');
++      return new SockJS('https://localhost/signaling');
+    }
+ }
+{% endhighlight %}
+
+I teoretycznie to wystarczy do działania, chociaż w celu zostawienia kompatybilności wstecz 
+oraz rozwiązania problemów podczas których wykorzystujemy SockJS w bibliotece wybudowanej za pomocą RollUp,
+konieczne jest zmodyfikowanie kodu. 
+
+W przypadku kompatybilności wstecz, został dodany dodatkowy endpoint wykorzystujący wersjonowanie.
+Problemy z RollUp zostały rozwiązane poprzez załadowanie biblioteki SockJS już po stronie klienta.
+
+
+### Wersja po uwzględnieniu poprawek
+#### Backend
+pom.xml
+{% highlight diff %}
+<dependencies>
+    <dependency>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-websocket</artifactId>
+    </dependency>
++   <dependency> <!-- własna wersja sockjs jest opcjonalna :) -->
++       <groupId>org.webjars.npm</groupId>
++       <artifactId>sockjs-client</artifactId>
++       <version>1.4.0</version>
++   </dependency>
+</dependencies>
+{% endhighlight %}
+
+WebSocketConfig.java
+{% highlight diff %}
+@Configuration
+@EnableWebSocket
++@EnableWebMvc
+-public class WebSocketConfig implements WebSocketConfigurer {
++public class WebSocketConfig implements WebSocketConfigurer, WebMvcConfigurer {
+    private SocketHandler socketHandler;
+
+    public WebSocketConfig(SocketHandler socketHandler) {
+        this.socketHandler = socketHandler;
+    }
+
+    @Override
+    public void registerWebSocketHandlers(WebSocketHandlerRegistry registry) {
+-       registry.addHandler(socketHandler, "/signaling");
++       registry.addHandler(socketHandler, "/signaling"); // pozostawione dla kompatybilności wstecz :)
++       registry.addHandler(socketHandler, "/v2/signaling/sockjs-server") // endpoint dla SockJS
++               .withSockJS()
++               // SockJS wykorzystuje iframe w specyficznych przypadkach
++               // i tutaj "nadmiarowo" podajemy własną wersję biblioteki JS z własnego serwera,
++               // zamiast https://cdn.jsdelivr.net/sockjs/1.0.0/sockjs.min.js
++               .setClientLibraryUrl("/v2/signaling/sockjs-static/sockjs-client/sockjs.min.js");
+    }
++   @Override
++   public void addResourceHandlers(ResourceHandlerRegistry registry) {
++       // tutaj również nadmiarowo wystawiamy statyczny zasób z biblioteką SockJS
++       registry.addResourceHandler("/v2/signaling/sockjs-static/sockjs-client/*.min.js")
++               .addResourceLocations("classpath:/META-INF/resources/webjars/sockjs-client/1.4.0/dist/")
```

Comment:
nie powiedziałbym żeby webjars to były state-of-the-art i coś czym chcemy się chwalić ;-)